### ISSUE TITLE
Compare --benchmark-ledger-ops to measurements of the actual node

### ISSUE THREAD

**Issue by nfrisby** [IntersectMBO/ouroboros-consensus#1272]
So far, we have assumed that `--benchmark-ledge-ops` reports relevant elapsed times. Issue https://github.com/IntersectMBO/ouroboros-consensus/issues/223 for example identifies that as a risk.

While recently attempting to determine "necessary" bandwidth requirements for a syncing node, I realized that same risk, and began an experiment.

This Issue is to finish that experiment, archive its source, and summarize my findings in a technical report.

**Comment by nfrisby** [comment]
This plot is one of my initial results.

![image](https://github.com/user-attachments/assets/b511b0f3-fc5f-4141-968e-7df004b82bc1)

The `pipe-valid` interval is delimited by two tracer events that enclose little more than the invocation of the ledger rules. In my patched node for this experiment, those tracers emit a `UserMessage` in the GHC eventlog. My analysis removes (a conservative approximation of) the GC times from that interval, and the result is plotted. (The node is syncing from a single peer: `immdb-server` on the loopback `127.0.0.1`. I have observed the same run-times when the node's db was mounted on a `ramfs`, so I don't think disk I/O is a disruption here.)

The `db-analyser` pass itself is single-threaded and uses the `GHC.RTSStats` API to access the mutator time, so both data points ought to just be the time spent in the ledger rules.

I am surprised that the node seems _faster_ than the `db-analyser` pass. (Both were run on the same machine, using the same dependencies, etc --- though I'm unsure about the compiler version being the same.)

The difference is on the order of 5ms, whereas the totals tend to the range 20ms to 40ms. So that's a significant disagreement, especially between already-smoothed data (smoothing the actual differences gives very similar results). The relative gap is even more significant in Byron (not shown).

**Comment by nfrisby** [comment]
I neglected to mention that it's well-correlated.
![image](https://github.com/user-attachments/assets/b4570b15-1cee-4783-b8d7-d7ce32d6c68e)

The regression line's slope is `1.18345`.

**Comment by nfrisby** [comment]
I've run the big sync twice now. Each involves ~1040 samples. 730 of those had the same slot (I do not yet have a confident explanation for sequence of queue pops differs at all!), and the scatter of the paired durations show strong linear correlation.

![image](https://github.com/user-attachments/assets/ffd3e38d-acd4-4a3b-a6ae-c794f20e8cbf)

**Comment by nfrisby** [comment]
I have confirmed that some blocks are being downloaded twice. I don't yet know why.

**Comment by nfrisby** [comment]
Having identified the reason runs were sampling different blocks, I was able to adapt the sampling logic to not be disrupted (ie to notice when a block was a duplicate and not count it towards the sampling period (which is 100 blocks)).

I then did 3 runs from genesis to slot 100million over the weekend; each run's samples indeed had the exact same slots. Here's the plot of the highest pipe-valid for each block minus the least pipe-valid.

![image](https://github.com/user-attachments/assets/f644e205-9447-47d1-b621-72159bad9d63)

![image](https://github.com/user-attachments/assets/90f4efaf-6b44-467c-9c17-7372d59ba764)

![image](https://github.com/user-attachments/assets/748447b5-055e-49b4-94b6-b9f0b622e99a)


(Because the sampling period is 100 blocks, x=45000 is approximately the end of Byron on this plot.)

**Comment by nfrisby** [comment]
If I normalize by the median, the three intervals still remain. In fact, the breakpoints are seem even easier to see.

![image](https://github.com/user-attachments/assets/39be2109-50d2-41f9-95f8-27e540faf068)

![image](https://github.com/user-attachments/assets/b6a825ae-599c-442c-beb1-ecbadc47746a)

![image](https://github.com/user-attachments/assets/bd4ee910-d404-4b9f-abf7-f783773f1f11)

I chatted with the Networking team, and Karl suggested that the second breakpoint (the first is obviously Byron->Shelley) coincides with the "SundaeSwap launch", during which the chain was congested for several weeks. (And during the first portion of it, most blocks were full.)

![image](https://github.com/user-attachments/assets/23c10d54-2a0f-4037-8800-5b7273661a44)

In other words, that's the first arrival of serious dApp load. And it seems it has remained.

**Comment by nfrisby** [comment]
The second changepoint is less apparent in the `db-analyser --benchmark-ledger-ops` data.

![image](https://github.com/user-attachments/assets/cc7bf45b-accc-4f32-b610-f61e86dabbd7)

![image](https://github.com/user-attachments/assets/f184703e-7311-4f69-8252-9bce1707336b)

(This data is recorded in microseconds instead of nanoseconds. As such, some values are actually equivalent. In that case, I nudge the 0 difference to 1 microsecond; that's why there's a multitude of values near the bottom of the plot.

---

Here's the same `db-analyser` data, without normalizing by the median of the triplicate measurements.

![image](https://github.com/user-attachments/assets/c241f3e3-153f-41de-b8f5-8e494cd66696)

![image](https://github.com/user-attachments/assets/34cdcafa-1757-483b-a5b0-9a531ca0e102)

**Comment by nfrisby** [comment]
The median of the node measurements is a pleasing match with the median of the `db-analyser` measurements (bottom plot in previous comment). Many outliers line up.

![image](https://github.com/user-attachments/assets/dd22bd93-9748-4136-9612-60dc1ce573f3)

**Comment by nfrisby** [comment]
The x-y scatter of the median of `pipe-valid` and the median of `--benchmark-ledger-ops` looks predominantly 1-to-1, but with `--benchmark-ledger-ops` being slightly faster.

![image](https://github.com/user-attachments/assets/6482d048-16b3-4f46-855f-94a7a73b6501)

**Comment by nfrisby** [comment]
I think this exercise is complete. See the report here.

https://docs.google.com/document/d/17Gps7v0w6yRHSMbLv3JkFsBX9TdqOfG-KRShd4eqEBU/edit?usp=sharing
